### ISSUE TITLE
Fixing PSR2 multiline call leaves first argument intact

### ISSUE THREAD

**Issue by VasekPurchart** [squizlabs/PHP_CodeSniffer#577]
Currently when fixing `PSR2.Methods.FunctionCallSignature.MultipleArguments` the first argument is left intact, which I believe does not correspond with the definition in [PSR2](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md#46-method-and-function-calls):

> Argument lists MAY be split across multiple lines, where each subsequent line is indented once. When doing so, the first item in the list MUST be on the next line, and there MUST be only one argument per line.

Code:
```php
<?php

foo(1, 2,
	3
);
```
Fixed code:
```php
<?php

foo(1,
	2,
	3
);
```
Expected fixed result:
```php
<?php

foo(
	1,
	2,
	3
);
```

Also the resulting fixed code is not considered an error, which I believe is a bug for the same reason.

**Comment by gsherwood** [comment]
When I run PHPCBF over the code, my fixed code looks like this:
```php
<?php

foo(
    1,
    2,
    3
);
```

Are you using the latest version of PHPCS?

**Comment by VasekPurchart** [comment]
It was tested on the 2.3.1 version, but that's not the issue (master/2.3.2) does the same.

I've done more digging and found out the reason - in my ruleset I exclude the `PSR2.Methods.FunctionCallSignature.ContentAfterOpenBracket` (because I want to allow inline comments at this place). So I guess the multiline fix actually only reorders the subsequent parameters and relies on `ContentAfterOpenBracket` to move the content to the next line?

**Comment by gsherwood** [comment]
> So I guess the multiline fix actually only reorders the subsequent parameters and relies on ContentAfterOpenBracket to move the content to the next line?

Yes. The way PHPCBF works is that each error code can both report and error and fix an error.

If you want a fix to be included in PHPCBF only, but not produce any errors when running PHPCS normally, you can change your ruleset to selectively apply the rule: https://github.com/squizlabs/PHP_CodeSniffer/wiki/Annotated-ruleset.xml#selectively-applying-rules

In this case, you'd write do something like this:
```xml
<rule ref="PSR2">
    <exclude phpcs-only="true" name="PSR2.Methods.FunctionCallSignature.ContentAfterOpenBracket"/>
</rule>
```
or
```xml
<rule ref="PSR2.Methods.FunctionCallSignature.ContentAfterOpenBracket">
    <severity phpcs-only="true">0</severity>
</rule>
```
Depending on how you've chosen to exclude the check in your ruleset.

**Comment by VasekPurchart** [comment]
Thanks, I haven't noticed this possibility to apply the rules selectively yet.

But I am afraid it still does not solve these types of situations, because if I exclude this only in phpcs, then the fixer will still convert this code (which is valid):

```php
<?php

foo( // foo
	1,
	2,
	3
);
```
into this:
```php
<?php

foo(
	// foo
	1,
	2,
	3
);
```
which is weird and I think semantically incorrect (because the comment belonged to the function call, like this it looks like a comment for the 1st argument maybe). So the code gets changed even when there was no "error" - I thought the phpcs only exclude would cause the rules be used in phpcbf only during the process of fixing (meaning that it would ignore this situation, but be used in the original one). Because like this I don't really see the point in using the selective rules (meaning you get no errors, but if you run phpcbf the code gets changed).

The point I was trying to make in #578 

> Are the fixes intended to work correctly even if some parts are excluded like this? I've quite a lot of other excludes, because I try to use as much as possible from existing sniffs before writing my own, but that means I differ in the details.

is that I think it would be more suitable to think of the errors when fixing as units and try to fix the errors they report completely. So if I have an `MultipleArguments` error I expect the fixer to fix the code completely.

But in this particular situation this is probably an wider issue, because code looking like this:

```php
<?php

foo(1,
    2,
    3
);
```
is not reported with `MultipleArguments`, although clearly not adhering to the PSR's

> Argument lists MAY be split across multiple lines, where each subsequent line is indented once. When doing so, the first item in the list MUST be on the next line, and there MUST be only one argument per line.

I know it's my fault in excluding part of the sniff, but the point I am trying to make is if the errors are treated more like units (`MultipleArguments` here covering the part I mention above) the existing sniffs would be far more reusable.

**Comment by gsherwood** [comment]
> So the code gets changed even when there was no "error" - I thought the phpcs only exclude would cause the rules be used in phpcbf only during the process of fixing (meaning that it would ignore this situation, but be used in the original one).

When you say a rule is phpcbf-only, you are saying that the rule should be used when running the phpcbf command and not when running the phpcs command. This means, the error is not reported, but it will be treated as an error by the auto-fixer and fixed anyway.

When you say a rule is phpcs-only, you are saying that you want the error reported, but you don't want the fixer to correct it.

If you don't want the error reported, and you don't want it fixed, exclude it completely.

> is not reported with MultipleArguments, although clearly not adhering to the PSR's

So re-enable the check that enforces the parenthesis as the last content on the line. If you really want comments after the opening parenthesis, you'll need to write your own sniff, ignore the errors when you get them, or start placing your comments above the function call.

I'm going to close this issue because everything is working as expected, although I know it's not working entire the way you want. I just don't have a solution for you that I can personally implement.

Feel free to keep chatting on this issue if you want help though, or have any suggestions.
